In [ ]:
import pandas as pd

sales = pd.read_csv("/sales_transactions.csv")

sales.info()

sales.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10589 entries, 0 to 10588
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   SaleID         10589 non-null  int64  
 1   Date           10577 non-null  object 
 2   BranchID       10589 non-null  object 
 3   ProductID      10589 non-null  object 
 4   CustomerID     5814 non-null   object 
 5   Quantity       10589 non-null  int64  
 6   UnitPrice      10589 non-null  int64  
 7   Amount         10589 non-null  float64
 8   PaymentMethod  10589 non-null  object 
dtypes: float64(1), int64(3), object(5)
memory usage: 744.7+ KB


,SaleID,Date,BranchID,ProductID,CustomerID,Quantity,UnitPrice,Amount,PaymentMethod
0,109375,5/1/2026,B05,P02,NaN,4,165,660.00,GCash
1,107239,1/18/2026,B01,P06,C0068,1,150,150.00,Cash
2,102371,5/11/2025,B06,P08,C0073,4,95,380.00,Cash
3,104038,8/14/2025,B04,P09,NaN,2,75,150.00,GCash
4,103925,8/9/2025,B02,P02,C0048,1,165,165.00,Cash
5,105769,11/12/2025,B04,P02,NaN,1,165,165.00,Cash
6,104150,8/20/2025,B06,P07,C0090,2,180,360.00,Cash
7,109794,5/22/2026,B01,P12,C0078,1,190,190.00,GCash
8,109394,5/2/2026,B04,P08,C0082,1,95,61.02,Cash
9,105161,10/13/2025,B06,P08,NaN,1,95,95.00,Cash


**Output**- I used gpt to be able to code it the right way and after the result i noticed the salesID is data type is recognised as int64 and asked gpt to change it before going to the next step.

In [ ]:
import pandas as pd

sales = pd.read_csv("/sales_transactions.csv")

sales["SalesID"] = sales["SalesID"].astype("string")

sales.info()

sales.head(10)

KeyError: 'SalesID'

**Output**- in here i noticed that the "SalesID" is supposed to be SaleID only so i changed it.

In [ ]:
import pandas as pd

sales = pd.read_csv("/sales_transactions.csv")

print(sales.columns.tolist())

['SaleID', 'Date', 'BranchID', 'ProductID', 'CustomerID', 'Quantity', 'UnitPrice', 'Amount', 'PaymentMethod']


In [ ]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10589 entries, 0 to 10588
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   SaleID         10589 non-null  int64  
 1   Date           10577 non-null  object 
 2   BranchID       10589 non-null  object 
 3   ProductID      10589 non-null  object 
 4   CustomerID     5814 non-null   object 
 5   Quantity       10589 non-null  int64  
 6   UnitPrice      10589 non-null  int64  
 7   Amount         10589 non-null  float64
 8   PaymentMethod  10589 non-null  object 
dtypes: float64(1), int64(3), object(5)
memory usage: 744.7+ KB


In [ ]:
import pandas as pd

sales = pd.read_csv("/sales_transactions.csv")

sales["SaleID"] = sales["SaleID"].astype("object")

sales.info()

sales.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10589 entries, 0 to 10588
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   SaleID         10589 non-null  object 
 1   Date           10577 non-null  object 
 2   BranchID       10589 non-null  object 
 3   ProductID      10589 non-null  object 
 4   CustomerID     5814 non-null   object 
 5   Quantity       10589 non-null  int64  
 6   UnitPrice      10589 non-null  int64  
 7   Amount         10589 non-null  float64
 8   PaymentMethod  10589 non-null  object 
dtypes: float64(1), int64(2), object(6)
memory usage: 744.7+ KB


,SaleID,Date,BranchID,ProductID,CustomerID,Quantity,UnitPrice,Amount,PaymentMethod
0,109375,5/1/2026,B05,P02,NaN,4,165,660.00,GCash
1,107239,1/18/2026,B01,P06,C0068,1,150,150.00,Cash
2,102371,5/11/2025,B06,P08,C0073,4,95,380.00,Cash
3,104038,8/14/2025,B04,P09,NaN,2,75,150.00,GCash
4,103925,8/9/2025,B02,P02,C0048,1,165,165.00,Cash
5,105769,11/12/2025,B04,P02,NaN,1,165,165.00,Cash
6,104150,8/20/2025,B06,P07,C0090,2,180,360.00,Cash
7,109794,5/22/2026,B01,P12,C0078,1,190,190.00,GCash
8,109394,5/2/2026,B04,P08,C0082,1,95,61.02,Cash
9,105161,10/13/2025,B06,P08,NaN,1,95,95.00,Cash


**Output**- After changing the salesID this is the result. This data shows that there are  10,589 entries/rows and 9 column.

In [ ]:
import pandas as pd

sales = pd.read_csv("/sales_transactions.csv")

sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10589 entries, 0 to 10588
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   SaleID         10589 non-null  int64  
 1   Date           10577 non-null  object 
 2   BranchID       10589 non-null  object 
 3   ProductID      10589 non-null  object 
 4   CustomerID     5814 non-null   object 
 5   Quantity       10589 non-null  int64  
 6   UnitPrice      10589 non-null  int64  
 7   Amount         10589 non-null  float64
 8   PaymentMethod  10589 non-null  object 
dtypes: float64(1), int64(3), object(5)
memory usage: 744.7+ KB


In [ ]:
sales.groupby("BranchID")["Amount"].sum().sort_values(ascending=False)

,Amount
BranchID,
B01,607298.18
B02,505149.08
B03,421895.03
B05,408700.00
B04,352604.76
B06,214558.75
b01,2750.00
B03,2325.00
B-05,1400.00


**Output**- As the results show that the B01 having the most total sales of 607,298.18 and B06 having only 214,558.75. But i also noticed that every branches have also an incosistent format like the b01.


In [ ]:
import pandas as pd

branches = pd.read_csv("/branches.csv")

print(branches.head())
print(branches.columns.tolist())

  BranchID  BranchName             City Province OpenedDate  Unnamed: 5  \
0      B01     Calamba     Calamba City   Laguna   3/1/2019         NaN   
1      B02  Santa Rosa  Santa Rosa City   Laguna  7/15/2020         NaN   
2      B03       Binan       Binan City   Laguna  1/20/2021         NaN   
3      B04   San Pablo   San Pablo City   Laguna  5/10/2022         NaN   
4      B05   Los Banos        Los Banos   Laguna   8/1/2023         NaN   

   Unnamed: 6  Unnamed: 7  Unnamed: 8  Unnamed: 9 Unnamed: 10 Unnamed: 11  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN        DAY        SALES   

  Unnamed: 12  Unnamed: 13  Unnamed: 14 Unnamed: 15 Unnamed: 16   Unnamed: 17  
0     

In [ ]:
merged = pd.merge(
    sales,
    branches,
    on="BranchID",
    how="inner"
)

len(sales), len(merged)

(10589, 10549)

**Output**- Result shows that after merging the datas it show there are 40 rows that are lost, having different formats to the data caused this.

In [ ]:
def validate(df, name):

    print(f"=== {name} ===")
    print("rows:", len(df))
    print("dtypes:\n", df.dtypes)
    print("missing:\n", df.isna().sum())
    print("duplicate ids:", df["SaleID"].duplicated().sum())
    print("negative amounts:", (df["Amount"] < 0).sum())

    orphans = set(df["BranchID"]) - set(branches["BranchID"])
    print("orphan BranchIDs:", orphans)

In [ ]:
validate(sales, "Sales")

=== Sales ===
rows: 10589
dtypes:
 SaleID             int64
Date              object
BranchID          object
ProductID         object
CustomerID        object
Quantity           int64
UnitPrice          int64
Amount           float64
PaymentMethod     object
dtype: object
missing:
 SaleID              0
Date               12
BranchID            0
ProductID           0
CustomerID       4775
Quantity            0
UnitPrice           0
Amount              0
PaymentMethod       0
dtype: int64
duplicate ids: 23
negative amounts: 0
orphan BranchIDs: {'B-02', 'B-05', 'b01', 'B06 ', 'b04', 'B03 '}


**Output**- The Sales table has 23 duplicate SaleIDs, which is a uniqueness problem. It also has 4,775 missing CustomerIDs, meaning many sales do not have a customer recorded. There are also 6 orphan BranchIDs caused by inconsistent formatting, such as lowercase letters, extra spaces, and hyphens. No negative amounts were found.

In [ ]:
sales["Date"] = pd.to_datetime(sales["Date"], errors="coerce")

In [ ]:
sales["Month"] = sales["Date"].dt.to_period("M")

In [ ]:
sales["BranchID"] = (
    sales["BranchID"]
    .astype(str)
    .str.strip()
    .str.upper()
    .str.replace("-", "", regex=False)
)

branches["BranchID"] = (
    branches["BranchID"]
    .astype(str)
    .str.strip()
    .str.upper()
    .str.replace("-", "", regex=False)
)

In [ ]:
merged = pd.merge(
    sales,
    branches,
    on="BranchID",
    how="inner"
)

In [ ]:
branch_month_revenue = (
    merged
    .groupby(["BranchID", "Month"])["Amount"]
    .sum()
    .reset_index()
)

In [ ]:
print("Rows:", len(branch_month_revenue))

Rows: 108


In [ ]:
branch_month_revenue.head(10)

,BranchID,Month,Amount
0,B01,2025-01,32340.00
1,B01,2025-02,27375.00
2,B01,2025-03,32480.00
3,B01,2025-04,31215.00
4,B01,2025-05,33165.00
5,B01,2025-06,30325.00
6,B01,2025-07,28815.00
7,B01,2025-08,40306.66
8,B01,2025-09,32560.00
9,B01,2025-10,28055.00


In [ ]:
print("Sales BranchIDs:")
print(sorted(sales["BranchID"].unique()))

print("\nBranches BranchIDs:")
print(sorted(branches["BranchID"].unique()))

print("\nMonths:")
print(sorted(sales["Month"].dropna().unique()))

Sales BranchIDs:
['B01', 'B02', 'B03', 'B04', 'B05', 'B06']

Branches BranchIDs:
['B01', 'B02', 'B03', 'B04', 'B05', 'B06', 'NAN']

Months:
[Period('2025-01', 'M'), Period('2025-02', 'M'), Period('2025-03', 'M'), Period('2025-04', 'M'), Period('2025-05', 'M'), Period('2025-06', 'M'), Period('2025-07', 'M'), Period('2025-08', 'M'), Period('2025-09', 'M'), Period('2025-10', 'M'), Period('2025-11', 'M'), Period('2025-12', 'M'), Period('2026-01', 'M'), Period('2026-02', 'M'), Period('2026-03', 'M'), Period('2026-04', 'M'), Period('2026-05', 'M'), Period('2026-06', 'M')]


**Output**- As for this part i double checked my rows in my table resulting of not the same result as my p1 say 143, now it say's 108 row having only 6 branches in 18 months totaling of 108 rows.

**C4 LAWFUL BASIS STATEMENT** - This slice uses sales transaction data containing BranchID, Date, and Amount to calculate total revenue by branch and month. It does not identify individual customers because the results are aggregated at the branch-month level and do not include customer names or CustomerID. Aggregating the data helps protect individual customer information because individual transactions are not shown in the final output. The analysis is intended only for business analysis and reporting purposes.